### PR TITLE
[SPARK-3181] [ML] Implement RobustRegression with huber loss.

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

The current implementation is a straight forward porting for Python scikit-learn `HuberRegressor`, so it produces the same result with that.
The code is used for discussion and please overpass trivial issues now, since I think we may have slightly different idea for our Spark implementation.

Here I listed some major issues should be discussed:
### Objective function.

We use Eq.(6) in [A robust hybrid of lasso and ridge regression](http://statweb.stanford.edu/~owen/reports/hhu.pdf) as the objective function.
image
But the convention is different from other Spark ML code such as `LinearRegression` in two aspects:
1, The loss is total loss rather than mean loss. We use `lossSum/weightSum` as the mean loss in `LinearRegression`.
2, We do not multiply the loss function and L2 regularization by 1/2. This is not a problem since it does not affect the result if we multiply the whole formula by a factor.
So should we turn to use the modified objective function like following which will be consistent with other Spark ML code?
image
### Implement a new class RobustRegression or a new loss function for LinearRegression？

Both `LinearRegression` and `RobustRegression` accomplish the same goal, but the output of `fit` will be different: `LinearRegressionModel` and `RobustRegressionModel`. The former only contains `coefficients`, `intercept`; but the latter contains `coefficients`, `intercept`, `scale/sigma` (and even the outlier samples similar to sklearn `HuberRegressor.outliers_`). It will also involve save/load compatibility issue if we combine the two models become one. One trick method is we can drop `scale/sigma` and make the `fit` by this huber cost function still output `LinearRegressionModel`, but I don't think it's an appropriate way since it will miss some model attributes. So I implemented `RobustRegression` in a new class, and we can port this loss function to `LinearRegression` if needed at later time. 
### Bugs of breeze LBFGS-B and work around.

The estimated parameter `\sigma` must > 0 which is a bound optimize problem and we should use `LBFGS-B` to solve, but there is a bug in breeze `LBFGS-B`. We figure out the work around with modified `LBFGS`. 
Since we known the huber loss function is convex in space `\sigma > 0` and the bound `\sigma = 0` is unreachable. The solution of loss function will not be on the bound. We still optimize the loss function by `LBFGS` but limit the step size when doing line search of each iteration. We should verify the step size generated by line search in the space `\sigma > 0`.
We override the function `LBFGS.determineStepSize` to limit the step size. We should make sure that `\sigma > 0` after take step operation: `x(k+1) = x(k) + alpha * dir` (`\sigma` is the first element of parameter vector in my implementation). We use `BacktrackingLineSearch` to do line search since it can be set the upper bound of the returned step size. Meanwhile, `BacktrackingLineSearch` still checks the strong wolfe conditions.
## How was this patch tested?

Unit tests.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+ * @param featuresStd The standard deviation values of the features.
+ * @param m The shape parameter to control the amount of robustness.
+ */
+private class HuberAggregator(
+    parameters: Vector,
+    fitIntercept: Boolean,
+    featuresStd: Array[Double],
+    m: Double) extends Serializable {
+
+  private val coefficients: Array[Double] = parameters.toArray.slice(2, parameters.size)
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: This aggregator will serialize the `featuresStd` and the `coefficients` between aggregation steps, which is not necessary. You can mark them as `@transient` or simply pass them to the add function as `LogisticRegression` does. You can see https://github.com/apache/spark/pull/14109 for more details.

**Comment by WeichenXu123** [comment]
@yanboliang I go through the code and there are several problems need to solve:

The robust regression has a parameter `sigma` which must > 0, so that it is a bound optimize problem and should use LBFGS-B. But as my test, current breeze LBFGS-B has bugs and when iterating sometimes it will generate NaN value and corrupt the computing. 

I add some log printing to help debug, I paste a small fragment to show how the LBFGS-B corrupt:
**(robust regression w/o intercept w/ regularization test)**
costFun: sigma param: 1.0
huberAggrLoss + reg: 18262.68068379334
cost grad- sigma: -630.1789355384457
costFun: sigma param: 631.1789355384457
huberAggrLoss + reg: 1.256602668595641E7
cost grad- sigma: -466.0711286869664
costFun: sigma param: 64.01789355384457
huberAggrLoss + reg: 483796.45119015244
cost grad- sigma: -448.1113667824356
costFun: sigma param: 9.849995439060637
huberAggrLoss + reg: 44154.79971484518
cost grad- sigma: -275.5999029061156
costFun: sigma param: 3.2447088269560513
huberAggrLoss + reg: 8513.171279631315
cost grad- sigma: -5.737776191290681
**costFun: sigma param: NaN
huberAggrLoss + reg: NaN**
cost grad- sigma: -822.4999999999944

as shown above, when sigma param became NaN in iterating, the LBFGS-B has corrupted and there is no need to continue. 

When I trace the LBFGS-B I found that in
`LBFGSB.subspaceMinimization` method, it may cause output point became (NaN, NaN...) even if the input is OK. so that I think it is a bug in `LBFGSB.subspaceMinimization` . 
I think this problem has no wark-around way and need Breeze community to fix it.

The second problem, whether the loss should divided by N and whether L2 reg should divided by 2, I think it should keep consistent with other GLM alogrithm in mllib.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+ *
+ * This supports two types of regularization: None and L2.
+ *
+ * This estimator is different from the R implementation of Robust Regression
+ * ([[http://www.ats.ucla.edu/stat/r/dae/rreg.htm]]) because the R implementation does a
+ * weighted least squares implementation with weights given to each sample on the basis
+ * of how much the residual is greater than a certain threshold.
+ */
+@Since("2.1.0")
+class RobustRegression @Since("2.1.0") (@Since("2.1.0") override val uid: String)
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Are all `@Since` required? I'd think the one on line 82 would be enough.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+  extends Regressor[Vector, RobustRegression, RobustRegressionModel]
+  with RobustRegressionParams with Logging {
+
+  @Since("2.1.0")
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: I don't think you need `@Since` at every symbol in the class (that was `@Since` itself with the same annotation).

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+   */
+  @Since("2.1.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+
+  override protected def train(dataset: Dataset[_]): RobustRegressionModel = {
+    val numFeatures = dataset.select(col($(featuresCol))).first().getAs[Vector](0).size
+    val w = if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+
+    val instances: RDD[Instance] =
+      dataset.select(col($(labelCol)), w, col($(featuresCol))).rdd.map {
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Why don't you use `as[Instance]`?

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+      instances.treeAggregate(new MultivariateOnlineSummarizer)(seqOp, combOp)
+    }
+
+    val featuresStd = featuresSummarizer.variance.toArray.map(math.sqrt)
+    val regParamL2 = $(regParam)
+
+    val costFun = new HuberCostFun(instances, $(fitIntercept), $(standardization),
+      featuresStd, regParamL2, $(m))
+
+//    val lowerBounds = BDV[Double](Array.fill(numFeatures + 2)(Double.MinValue))
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Why do you need to keep these comments?

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+    if (!$(fitIntercept)) initialArray(1) = 0.0
+    val initialParameters = Vectors.dense(initialArray)
+
+    val states = optimizer.iterations(new CachedDiffFunction(costFun),
+      initialParameters.asBreeze.toDenseVector)
+
+    val parameters = {
+      val arrayBuilder = mutable.ArrayBuilder.make[Double]
+      var state: optimizer.State = null
+      while (states.hasNext && !(state != null && state.searchFailed)) {
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Could you `states.foreach` or similar with `filter`? Double negation begs for a simpler solution.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+
+  private var totalCnt: Long = 0L
+  private var weightSum: Double = 0.0
+  private var lossSum = 0.0
+  // Here we optimize loss function over sigma, intercept and coefficients
+  private val gradientSumArray = Array.ofDim[Double](parameters.size)
+
+  def count: Long = totalCnt
+  def loss: Double = {
+    require(weightSum > 0.0, s"The effective number of instances should be " +
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Why do you `require` against `weightSum` here?

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+      s"greater than 0.0, but $weightSum.")
+    lossSum
+  }
+  def gradient: Vector = {
+    require(weightSum > 0.0, s"The effective number of instances should be " +
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Why do you `require` against `weightSum` here?

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+      }
+
+      totalCnt += 1
+      weightSum += weight
+      this
+    }
+  }
+
+  /**
+   * Merge another HuberAggregator, and update the loss and gradient
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: `[[HuberAggregator]]` to make a reference?

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+
+    if (other.weightSum != 0) {
+      totalCnt += other.totalCnt
+      weightSum += other.weightSum
+      lossSum += other.lossSum
+
+      var i = 0
+      val localThisGradientSumArray = this.gradientSumArray
+      val localOtherGradientSumArray = other.gradientSumArray
+      while (i < gradientSumArray.length) {
```
Comment (jaceklaskowski) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Why not `gradientSumArray.map`?

Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Although this will be called far less than the `add` method, I believe we still consider this performance critical. A while loop will be faster and require no virtual function calls compared with a `foreach`. We don't use `map` because we are modifying in place here.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+
+    val costFun = new HuberCostFun(instances, $(fitIntercept), $(standardization),
+      featuresStd, regParamL2, $(m))
+
+//    val lowerBounds = BDV[Double](Array.fill(numFeatures + 2)(Double.MinValue))
+//    lowerBounds(0) = 1E-20
+//    val upperBounds = BDV[Double](Array.fill(numFeatures + 2)(Double.MaxValue))
+
+//    val optimizer = new BreezeLBFGSB(lowerBounds, upperBounds, $(maxIter), 10, $(tol))
+    val optimizer = new BreezeLBFGS[BDV[Double]]($(maxIter), 10, $(tol))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: See ```Potential bugs of breeze LBFGS-B``` section of the PR description.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala (hunk 5 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Please refer ```Potential bugs of breeze LBFGS-B``` section of the PR description. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Very good suggestion. Thanks!

**Comment by yanboliang** [comment]
cc @dbtsai @MechCoder

**Comment by dbtsai** [comment]
I'm making through the first pass now.

**Comment by tewf** [comment]
Could we instead implement a more general Robust Linear Model [M-estimator](http://research.microsoft.com/en-us/um/people/zhang/INRIA/Publis/Tutorial-Estim/node24.html) type like is done in [statsmodels RLM](http://statsmodels.sourceforge.net/0.6.0/rlm.html), see [RLM.py](http://statsmodels.sourceforge.net/0.6.0/_modules/statsmodels/robust/robust_linear_model.html#RLM)?  The Huber loss would then be one of the M-estimators, maybe the default as done in statsmodels.

I think that the [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala) was made and intended to aid in developing a robust M-Estimator framework.

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+
+  /**
+   * The shape parameter to control the amount of robustness. Must be > 1.0.
+   * At larger values of M, the huber criterion becomes more similar to least squares regression;
+   * for small values of M, the criterion is more similar to L1 regression.
+   * Default is 1.35 to get as much robustness as possible while retaining
+   * 95% statistical efficiency for normally distributed data.
+   */
+  @Since("2.1.0")
+  final val m = new DoubleParam(this, "m", "The shape parameter to control the amount of " +
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Change @Since

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+     * Huber loss function should be optimized by LBFGS-B, but breeze LBFGS-B has bug and
+     * can't be used currently. We figure out one work around with modified LBFGS.
+     * Since we known huber loss function is convex in space "\sigma > 0" and the bound
+     * "\sigma = 0" is unreachable. The solution of loss function will not be on the bound.
+     * We still optimize the loss function by LBFGS but limit the step size when doing line search
+     * of each iteration. We should verify the step size generated by each line search in the
+     * space "\sigma > 0".
+     */
+    val optimizer = new BreezeLBFGS[BDV[Double]]($(maxIter), 10, $(tol)) {
+      override protected def determineStepSize(
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Update LBFGS-B as already fix bug in https://github.com/scalanlp/breeze/pull/633

File: mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala (hunk 10 above)
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala]: Not only function calling overhead, `Array.zipWithIndex` it will generate another temporary array.

**Comment by yanboliang** [comment]
@WeichenXu123 Thanks for your comment. I will update my PR ASAP.

**Comment by yanboliang** [comment]
I'll close this PR and open a new one. Feel free to review and comment. Thanks.

**Comment by yanboliang** [comment]
Please go to #19020 for reviewing and comments. Thanks.
